mersenne-31/src/quintic.rs

```diff
@@ -0,0 +1,96 @@
+use core::ops::Mul;
+
+use p3_field::{
+    define_ext, field_to_array, impl_ext_af, impl_ext_f, AbstractExtensionField, AbstractField,
+    Field,
+};
+
+use crate::Mersenne31;
+
+// F[X]/(X^5 - X - 6)
+define_ext!(pub M31x5([Mersenne31; 5]));
+
+impl<AF: AbstractField<F = Mersenne31>> Mul for M31x5<AF> {
+    type Output = Self;
+    #[rustfmt::skip]
+    fn mul(self, rhs: Self) -> Self::Output {
+        let m = |i: usize, j: usize| self.0[i].clone() * rhs.0[j].clone();
+        // prod = lo + hi * X^5
+        //      = lo + hi * (X + 6)
+        let lo = Self([
+            m(0,0),
+            m(1,0) + m(0,1),
+            m(2,0) + m(1,1) + m(0,2),
+            m(3,0) + m(2,1) + m(1,2) + m(0,3),
+            m(4,0) + m(3,1) + m(2,2) + m(1,3) + m(0,4),
+        ]);
+        let hi = Self([
+                     m(4,1) + m(3,2) + m(2,3) + m(1,4),
+                              m(4,2) + m(3,3) + m(2,4),
+                                       m(4,3) + m(3,4),
+                                                m(4,4),
+            AF::zero(),
+        ]);
+
+        // shift hi right to get hi * X
+        let mut hi_times_x = Self::zero();
+        hi_times_x.0[1..].clone_from_slice(&hi.0[..4]);
+
+        // maybe better reduction possible for * 6 ?
+        let w = AF::from_f(Mersenne31::from_canonical_u8(6));
+        lo + hi * w + hi_times_x
```

Comment:
This can probably be optimized to some extent (possibly by optimizing the polynomial too). No need to change it here but maybe leave a TODO or make an issue?